lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobDriver.java

```diff
@@ -230,41 +234,40 @@ private CLRBufferedLogHandler getCLRBufferedLogHandler() {
-      LOG.log(Level.INFO, "Allocated Evaluator: {0}, total running running {1}",
-          new Object[]{eval.getId(), JobDriver.this.contexts.size()});
-      if (JobDriver.this.handlerManager.getAllocatedEvaluatorHandler() == 0) {
-        throw new RuntimeException("Allocated Evaluator Handler not initialized by CLR.");
-      }
-      final AllocatedEvaluatorBridge allocatedEvaluatorBridge =
-          this.allocatedEvaluatorBridgeFactory.getAllocatedEvaluatorBridge(eval, this.nameServerInfo);
-      allocatedEvaluatorBridges.put(allocatedEvaluatorBridge.getId(), allocatedEvaluatorBridge);
-      NativeInterop.clrSystemAllocatedEvaluatorHandlerOnNext(
-          JobDriver.this.handlerManager.getAllocatedEvaluatorHandler(), allocatedEvaluatorBridge, this.interopLogger);
+    eval.setProcess(process);
+    LOG.log(Level.FINE, "Allocated Evaluator: {0}, total running running {1}, time: {2}",
+        new Object[]{eval.getId(), JobDriver.this.contexts.size(), dateFormat.format(new Date())});
+    final long handler = JobDriver.this.handlerManager.getAllocatedEvaluatorHandler();
+    if (handler == 0) {
+      throw new RuntimeException("Allocated Evaluator Handler not initialized by CLR.");
     }
+    final AllocatedEvaluatorBridge allocatedEvaluatorBridge =
+        this.allocatedEvaluatorBridgeFactory.getAllocatedEvaluatorBridge(eval, this.nameServerInfo);
+    allocatedEvaluatorBridges.put(allocatedEvaluatorBridge.getId(), allocatedEvaluatorBridge);
+    NativeInterop.clrSystemAllocatedEvaluatorHandlerOnNext(
+        handler, allocatedEvaluatorBridge, this.interopLogger, this.nameServerInfo, eval.getId());
+    LOG.log(Level.FINE, "End of JobDriver.Allocated Evaluator: {0}, time: {1}",
+        new Object[] {eval.getId(), dateFormat.format(new Date())});
   }
 
   private void handleFailedEvaluator(final FailedEvaluator eval, final boolean isRestartFailed) {
     try (final LoggingScope ls = loggingScopeFactory.evaluatorFailed(eval.getId())) {
-      synchronized (JobDriver.this) {
-        LOG.log(Level.SEVERE, "FailedEvaluator", eval);
-        for (final FailedContext failedContext : eval.getFailedContextList()) {
-          final String failedContextId = failedContext.getId();
-          LOG.log(Level.INFO, "removing context " + failedContextId + " from job driver contexts.");
-          JobDriver.this.contexts.remove(failedContextId);
-        }
-        String message = "Evaluator " + eval.getId() + " failed with message: "
-            + eval.getEvaluatorException().getMessage();
-        JobDriver.this.jobMessageObserver.sendMessageToClient(message.getBytes(StandardCharsets.UTF_8));
+      LOG.log(Level.SEVERE, "FailedEvaluator", eval);
+      for (final FailedContext failedContext : eval.getFailedContextList()) {
+        final String failedContextId = failedContext.getId();
+        LOG.log(Level.INFO, "removing context " + failedContextId + " from job driver contexts.");
+        JobDriver.this.contexts.remove(failedContextId);
```

Comment:
Maybe check the return value to make sure that the context was actually in the map.

Reply:
Romove will do nothing if doesn't exist.